cmd/syncthing: Conditionally enable CORS

letiemble opened the pull request:
### Purpose

This PR re-enables the "Access-Control-Allow-Origin" header so clients like browsers can perform CORS requests. This fixes issue #3540.

### Testing

New tests have been included.

AudriusButkevicius commented:
No, this allows to highjack the UI from any random URL.

letiemble commented:
Is it possible to re-enable this header for the /rest/* path ?

AudriusButkevicius commented:
That does not make it less of a security hole

letiemble commented:
What about making it configurable ?
- No value (by default), the header is not included
- A domain: this restricts CORS requests to a specific domain
- "*": this opens CORS requests to all domains

AudriusButkevicius commented:
Maybe, yet ideally I'd avoid that too as I bet there will be a guide somewhere on the internet explaining that you need to set it to '*' without explaining what the implications are, so it's very easy to make people slip.

What's the use-case here?

calmh commented:
Yes, lets talk about the use case. As we currently don't allow any REST access without either an API key or a CSRF token this could potentially be safe, even in the face of no authentication. Our previous issue was the combination of Access-Control-Allow-Origin, no authentication, and being able to GET the config without CSRF token.

letiemble commented:
Use-case
---------
I am working on a centralised dashboard to monitor all my Syncthing instances. The application is a Single Page Application written in JavaScript that runs into a browser. The application queries the Syncthing instances by making REST calls trough XHR.

In order for these calls to work, each Syncthing endpoint MUST:
- support pre-flight OPTIONS requests because of the X-API-KEY header
- provide a "Access-Control-Allow-Origin" header because the domain of the application is different from the domains of the Syncthing instances

---

My understanding is that as long as the REST API is fully protected by an API key or CSRF token, the "Access-Control-Allow-Origin" does not weaken these protections. I found [this post](http://stackoverflow.com/questions/25862467/cors-csrf-clarification) to be a nice summary on the topic.

For a better control, the "Access-Control-Allow-Origin" header can be made configurable:

1. No value (by default): the header is not included
1. An origin: this restricts CORS requests to the specified origin
1. "*": this opens CORS requests to all origins

For the two last values, a warning will be displayed in the GUI like when no user/password have been set for an open Syncthing instance.

Finally, I think the `/rest/system/config` endpoint should strip the `user`/`password`/`apiKey` information on GET (or any sensible information), while accepting them on POST.

calmh commented:
> - support pre-flight OPTIONS requests because of the X-API-KEY header
> - provide a "Access-Control-Allow-Origin" header because the domain of the application is different from the domains of the Synching instances

Would it be possible/viable to only set the A-C-A-O header when a valid X-API-Key is received? Or will the browser not add that to the OPTIONS request? For your use case, which is entirely valid IMO, we can always require an API key. For the "regular" GUI CSRF-token stuff I don't see a need to permit foreign origins.

> Finally, I think the /rest/system/config endpoint should strip the user/password/apiKey information on GET (or any sensible information), while accepting them on POST.

Yes. Well, the user isn't that sensitive and the password is bcrypt hashed, but the API key should not be visible in the config get. It just requires a rewiring of the GUI for that to happen.

AudriusButkevicius commented:
This looks acceptable.

letiemble commented:
Based on my tests (Chrome and Firefox), here is what I see:
- a GET request is initiated with a X-API-KEY header via XHR.
- the browser performs an OPTIONS preflight request (because of the X-API-KEY header). This request is made without the X-API-KEY header.
- the OPTIONS response MUST have the A-C-A-O header; otherwise the browser rejects the response.
- the browser then performs the actual GET request with the X-API-KEY.
- the GET response MUST have the A-C-A-O header; otherwise the browser rejects the response.

This behavior matchs the [specifications](https://www.w3.org/TR/cors/).

I have modified the PR in order to:
- set the A-C-A-O header for OPTIONS requests
- set the A-C-A-O header for GET/POST requests when a valid API key is provided.

letiemble commented:
Commits squashed.

aarredon commented:
Question. The last release v0.14.5 has an unverified signature. The GPG keys match up, but github reports that the email used doesn't match the committer email. Is this a cause for concern or just an oversight? Thanks and keep up all the awesome work!

AudriusButkevicius commented:
Please don't piggy back on a random pull request for questions like this.

calmh commented:
@st-review lgtm

Commit message needs tweaking, leaving final lgtm to Audrius

st-review commented:
@calmh: Noted! Need another LGTM or explicit merge command.

calmh commented:
@AudriusButkevicius any remaining concerns on this?

AudriusButkevicius commented:
@st-review lgtm

cmd/syncthing: Conditionally enable CORS

st-review commented:
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

calmh commented:
@st-review merge it

cmd/syncthing: Conditionally enable CORS

st-review commented:
@calmh: Merge failed:

```
$ git fetch -f origin refs/pull/3541/head:pr-3541
$ git fetch -f origin master:orig/master
From github.com:syncthing/syncthing
   49910a1..7990ffc  master     -> orig/master
$ git reset --hard
HEAD is now at 7990ffc cmd/syncthing: Copy config on upgrade, instead of renaming (fixes #3525)
$ git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.
$ git reset --hard orig/master
HEAD is now at 7990ffc cmd/syncthing: Copy config on upgrade, instead of renaming (fixes #3525)
$ git clean -fxd
Skipping repository syncthing/syncthing
Skipping repository syncthing/syncthing-android
Skipping repository syncthing/docs
Skipping repository syncthing/discosrv
Skipping repository syncthing/syncthing-gtk
Skipping repository syncthing/syncthing-inotify
$ git merge --squash --no-commit pr-3541
Auto-merging cmd/syncthing/gui_test.go
CONFLICT (content): Merge conflict in cmd/syncthing/gui_test.go
Auto-merging cmd/syncthing/gui.go
Squash commit -- not updating HEAD
Automatic merge failed; fix conflicts and then commit the result.

```

calmh commented:
@letiemble test needs a merge from master... :/

letiemble commented:
While testing `POST` requests, I noticed that the `Content-Type` header was missing from the `Access-Control-Allow-Headers` returned by OPTIONS. I have added it in the `corsMiddleware`.

I have rebased the changes on the latest `master` and change the commit message.

AudriusButkevicius commented:
@st-review merge it

cmd/syncthing: Conditionally enable CORS

st-review commented:
:ok_hand: Merged as 3990014073eb112933fe04c24bad564ce669c3bb. Thanks, @letiemble!
